### ISSUE TITLE
Voxelmorph implementation 

### ISSUE THREAD

**Issue by Taariq78** [voxelmorph/voxelmorph#216]
Hello, I hope you are well.  My name is Taariq Daniels; I am a final year Mechatronics Engineering student at the University of Cape Town that is currently busy with a research project.

My topic is 2d/3d medical image registration and I am attempting to implement an adaption of Voxelmorph but have had little success so far.  Would it be appropriate for me to ask some questions here and maybe receive some advice on the issues I am currently facing? 

Kind regards

**Comment by adalca** [comment]
@Taariq78 sure, ask away

Also, is your implementation not meeting your needs to start off, or are you doing this as an exercise?

**Comment by Taariq78** [comment]
Thank you for the response.
To answer your question (if I’m understanding it correctly), I am not doing this as an exercise.   Based on what I understand from your voxelmorph paper, it should be capable of achieving the main objectives of my research project.

I am fairly new to machine learning but please try to bear with me.
  
I am attempting to adapt voxelmorph, such that given an unseen 2-dimensional (grayscale) image pair, comprising of a fixed/target image and a moving image, the model will be able to align these images within the same coordinate frame.  The moving image is simply a displaced version of the fixed image (translation in x, y direction and/or rotated about the z-axis).
  
The current implementation of voxelmorph is designed to align 3-dimensional volumes; but I feel that this could be easily adapted to as how I described above.

I attempted to emulate the method as described within the paper, rather than look directly at the code you made available, as I could only follow to a limited extent and I did not wish to make use of code that I did not understand.
Correct if I am wrong:

- Voxelmorph uses a type of Unet architecture that, given an image pair, produces what you refer to as a flow field.

- This flow field is applied to the moving image and the difference between the target image and the warped image is propagated back as the error.

- Through training, the network will be able to infer a function that is able to produce a suitable flow field capable of aligning a given image pair.

My attempt of implementing this is as follows:

- I currently have a basic Unet implementation, which serves as my localisation network (forming part of the Spatial transformer network).  Unet: https://arxiv.org/pdf/1505.04597.pdf

- I made it such that the output of my network produces a flow field (or at least what I believe to be a flow field).  My implementation of the STN is based on this example: https://pytorch.org/tutorials/intermediate/spatial_transformer_tutorial.html

- I noticed that in your Pytorch (which is what I’m using) implementation that you do make use of the function torch.nn.functional.affine_grid(), which produces the desired flow fields.  Instead you use a different method of creating a sampling grid (screenshot attached).  Was this intentional?  
![sampling grid](https://user-images.githubusercontent.com/70844260/92376323-b3faad00-f102-11ea-915e-29f861e7d0fc.PNG)


- The flow field is then applied to my moving image using the torch.nn.functional.grid_sample() function, where the inputs are the moving image and the flow field (from the network)

- Using the MSE cost function (inputs being the moving and target images), I acquire the value of the similarity measurement between the image pair.  I’m using the Adam optimiser function.

- I chose to train my model on a small subset of my data to see if there are any signs of convergence before importing the rest of my data

- My training data is comprised of 2000 radiograph images (produced from CT scans):
     o	Consisting of 4 patients (500 images per patient)
     o	These images serve as the moving images that were produced by randomly displacing the original image 500 times.

- Training involves feeding a moving image with its corresponding fixed (original) image into the network (batch size of 5; 5 image pairs)

- How important is it to feed a random image pair into the network?  Currently, at each epoch, I randomise the order in which a patient is fed into the network.  E.g. epoch 1, patient 2 is first selected and every image pair corresponding to patient 2 is fed into the network.  After which, the next patient is selected, and all its pairs are fed into the network; this continues until each patient (and their corresponding pairs) have been fed into the network.  At the next epoch, the order is shuffled and the process repeats.        

At the time of writing this, my network shows no signs of learning.  I apologise for writing this much but any assistance/advice, as to where I might be going wrong would be greatly appreciated.  Thank you for your time.

**Comment by adalca** [comment]
Hi @Taariq78,

VoxelMorph is implemented for both 2D and 3D (in fact, also for 1D!) -- all of the core code works for any of these dimensions. It's true that in our papers we mostly work with 3D data, but we worked hard to make the implementation handle any dimension. In fact, the tutorial we created for people to learn about VoxelMorph (https://www.kaggle.com/adalca/learn2reg) registers 2D images (because it's faster so it works nicely in a tutorial). 

We've especially tested all of these things for the keras/tf version, but the pytorch version should also work for any dimension, and if it doesn't work for you please let us know. 

So I recommend you give all this a try. But to answer your other questions: I think all your steps sounds reasonable to me. I don't fully understand your `affine_grid` question -- are you only looking to use affine registration (not deformable?). And yes, randomizing the input training data is very useful!

**Comment by Taariq78** [comment]
@adalca  Okay, thank you, I'll make use of the tutorial.

My understanding is that affine registration is limited to translation and/or rotation?  As my data is comprised of images that are displaced as I described earlier, I don't believe I require a model capable of deformable registration.  So yes, I wish to use affine registration.  

For my question regarding the code snippet: what I was trying to understand is what was the purpose of not using the simple affine_grid() function?  Does this not produce the required flow field?  Or is it because voxelmorph is capable of deformable registration and therefore affine_grid() would not be adequate?  I hope this makes my question clearer

**Comment by Taariq78** [comment]
@adalca Hello again.  I'm attempting to train my data on the voxelmorph model provided but have ran into an issue.  I'm encountering this error in the grad_loss function.  I ran the individual components of the function and they seem to work fine; but the moment I try to use the function as a whole, I get the error as seen below.  Any idea as to why this might be happening?  Thank you. 

![grad_loss](https://user-images.githubusercontent.com/70844260/92483632-568d5b80-f1e9-11ea-9ef0-61f60d35cfc5.PNG)

**Comment by balakg** [comment]
Not quite sure why this is happening. Can you try printing out the size of variable 's' in the gradient_loss function?

**Comment by Taariq78** [comment]
Hello @balakg.  I re-initialised all of my code and it seems to be working now.  Thank you though.

Quick question: from looking through the pytorch implementation, I noticed that in the train.py file, where you rearrange the dimensions of the tensors using permute(), it assumes that the tensor is of 5 dimensions [(batch_size,channel,W,H,D)]?  I wish to train on 2-dimensional images (so the tensor is of 4 dimensions) and to do so, I had to modify the permute() line.  It's currently training but please let me know if what I did was actually
![permute](https://user-images.githubusercontent.com/70844260/92502280-ef2ed600-f1ff-11ea-8f42-c108ca8a1388.PNG)
 necessary?

**Comment by balakg** [comment]
What you did looks right to me, if you are working with 2D images. We did not yet add account for 2D in the code, apologies for that.

**Comment by adalca** [comment]
@balakg meant for the pytorch code -- but those two changes that @Taariq78 made should cover it, right?

**Comment by balakg** [comment]
Yes, I think so.
